### ISSUE TITLE
Question/Issue: Index not showing 

### ISSUE THREAD

**Issue by Bhersbach** [maximilianh/cellBrowser#251]
We are setting up Cell Browser in our lab and have successfully prepared two Cell Browser folders as a test. However, when we use cbBuild to host the data, we can access the individual datasets only by using the dataset name, i.e., server address:/?ds=dataset_name. The index/title page, does not load and, instead, we get an empty Cell Browser page. When attempting to open a dataset through file > open dataset, no options are given either. Is there something we are forgetting to specify? 

System specs:
OS: Linux (Ubuntu 18.04)
Cell Browser v1.2.1

**Comment by maximilianh** [comment]
Yes, I'm sure we can solve this quickly. Have you defined the dataset
hierarchy as explained here or not:
https://cellbrowser.readthedocs.io/en/develop/collections.html

Does your instance have a publicly accessible URL so I can have a quick
look?

**Comment by Bhersbach** [comment]
I hadn't done that and just gave it a try, it doesn't seem to have solved the issue (or I did it incorrectly). Unfortunately, our instance is currently not public, so I will try to briefly summarize the situation: 
- We have one folder (User1) that contains two datasets (Dataset1 & Dataset2)
- The ~/.cellbrowser.conf file does not seem to be present the home folder so I ran `export CBDATAROOT='/home/cellbrowser/User1/'`
- I copied cellbrowser.conf and desc.conf from Dataset1 to the User1 folder and edited as described in the link
- I run cbBuild -r /home/cellbrowser/html -p 8999 within the User1 folder, this runs but does not start a CellBrowser server
- I tried running the same command within the Dataset1 folder, which runs and creates a CellBrowser server but no index/title page

**Comment by maximilianh** [comment]
My ~/.cellbrowser.conf looks like this:

gaTag = "G-5CVR5ZC288" # test, GA3 and GA4 combined
htmlDir = "/usr/local/apache/htdocs-cells"
dataRoot = "/hive/data/inside/cells/datasets/"
outDirs = {"alpha" : "/usr/local/apache/htdocs-cells", "beta" :
"/usr/local/apache/htdocs-cells-beta/" }
reqTags = ["body_parts", "unit"] # you won't need this
email = ***@***.***"
#useTest = True # only for testing cbGenes
#onlyLower = True # only if you use the special DNS rules that we have
matrixFormat = "mtx" # default matrix format

**Comment by Bhersbach** [comment]
Thanks! My home directory doesn't have a .cellbrowser.conf file though. Is it located somewhere else when using a conda environment by any chance?

**Comment by maximilianh** [comment]
How about just creating a config file?

echo 'dataRoot = "/home/cellbrowser/User1/"' > ~/.cellbrowser.conf

I think I'll add this command to the documentation...

**Comment by Bhersbach** [comment]
I ran the code as you described. The webserver it launches is empty, I cannot access any datasets even by their url. Going to the html folder and opening the index html gives the following error: Could not find a dataset at dataset.json?86f083f58d. Would it make sense to do a video call so I show you?

**Comment by maximilianh** [comment]
We don't usually do video calls, also I don't know your time zone or
affiliation. We could do that, but as a last resort.

Yes it cannot find the dataset.json file.

Did you check ~/.cellbrowser.conf, what is in the file? Can you copy it
here?

Did you check this: "In  /home/cellbrowser/html there should be a desc.json
file then with your html text. and a dataset.json file."

**Comment by Bhersbach** [comment]
So our ~/.cellbrowser.conf file contains a single line: dataRoot = "/home/cellbrowser/OutputData/Bob/"
Within the html folder there are indeed the desc.json, the inded.html and dataset.json file

**Comment by maximilianh** [comment]
Previously you wrote that you used this command:

CBDATAROOT='/home/cellbrowser/User1/

But in the config file you have:

dataRoot = "/home/cellbrowser/OutputData/Bob/"

looks odd to me, are the data files of dataset1 and dataset2 really located
under /home/cellbrowser/OutputData/Bob/ ? So you have directories such
as  /home/cellbrowser/OutputData/Bob/dataset1
and  /home/cellbrowser/OutputData/Bob/dataset2?

And you have the top-level cellbrowser.conf and a desc.conf with a general
description of your overall webserver in the
directory "/home/cellbrowser/OutputData/Bob/" ?

**Comment by maximilianh** [comment]
The data root folder is determined in this code function:

def findRoot(inDir=None):
    """ return directory dataRoot defined in config file or CBDATAROOT
    environment variable.
    """

    if 'CBDATAROOT' in os.environ:
        dataRoot = os.environ['CBDATAROOT']
    else:
        dataRoot = getConfig("dataRoot")

    if dataRoot is None:
        logging.info("dataRoot is not set in ~/.cellbrowser.conf or via
$CBDATAROOT. Dataset hierarchies are not supported.")
        return None

    dataRoot = abspath(expanduser(dataRoot).rstrip("/"))

    if inDir is not None and not abspath(inDir).startswith(dataRoot):
        logging.info("input directory %s is not located under dataRoot %s.
Deactivating hierarchies." % (inDir, dataRoot))
        return None

    return dataRoot

Can you look for any of these messages in the cbBuild output and let me
know what they are?

**Comment by maximilianh** [comment]
More concretely, can you look for "Deactivating hierarchies" or "Dataset
hierarchies are not supported" ?

**Comment by tkakar** [comment]
Hi. I'm running into same proplem. In my case a computer restart worked and data hierarchy was created, at least inside the html folder I could see 'inDir' and the path to my root data directory in html/dataset.json file. However, when I open the browser by typing `python -m http.server 8080` inside the html directory, the UI shows nothing. Under the developer tools, I see the following error message. 

<img width="858" alt="Screen Shot 2022-09-15 at 2 39 40 PM" src="https://user-images.githubusercontent.com/14091615/190483910-f7889565-013e-4687-900f-ed7e0e73746b.png">

**Comment by maximilianh** [comment]
Ohhh!!! Darn. I can fix this. Yes, we have new "disease filters".
Apparently somewhere it assumes that "disease=["cancer"]" or something like
this is defined in the cellbrowser.conf file.

Are you using a github clone? If I fix this now, could you use the updated
source code form Github or do you need a pip release?

**Comment by tkakar** [comment]
Pip release please. Thank you.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Maximilian Haeussler ***@***.***>
Sent: Thursday, September 15, 2022 4:53:08 PM
To: maximilianh/cellBrowser ***@***.***>
Cc: Kakar, Tabassum ***@***.***>; Comment ***@***.***>
Subject: [EXT] Re: [maximilianh/cellBrowser] Question/Issue: Index not showing (Issue #251)


Ohhh!!! Darn. I can fix this. Yes, we have new "disease filters".
Apparently somewhere it assumes that "disease=["cancer"]" or something like
this is defined in the cellbrowser.conf file.

Are you using a github clone? If I fix this now, could you use the updated
source code form Github or do you need a pip release?

**Comment by maximilianh** [comment]
Hi @tkakar and @Bhersbach, sorry that it took me so long to understand what's going on. It had nothing to do with the hierarchy, but was a typo in the new diseases filter, I had fixed that in the develop branch, it slipped into the last release. Here is a new pre-release https://pypi.org/project/cellbrowser/1.2.1a1/  (I must release soon) You can install it with 
"pip install --pre cellbrowser --upgrade"

**Comment by tkakar** [comment]
@maximilianh  thanks for fixing the issue. I did as suggested. The `disease` related error is gone. 
The individual datasets can be accessed via `?ds=datasetName` on the url, but the data hierarchy still does not work. Under developer tools, I see errors. Posting screen shot of both scenarios (individual dataset and hiararchy).
<img width="1510" alt="Screen Shot 2022-09-16 at 10 49 42 AM" src="https://user-images.githubusercontent.com/14091615/190668367-485e6242-d24d-493b-b673-a55caecae6b3.png">
<img width="1302" alt="Screen Shot 2022-09-16 at 10 49 10 AM" src="https://user-images.githubusercontent.com/14091615/190668877-8b16bc15-8e72-48e6-81ac-01c9f45cd933.png">

**Comment by maximilianh** [comment]
Hmm....It would be so much easier to debug this if you had this on a public
website so I could run the debugger instead of guessing...

I have a feeling that we've fixed this bug in the develop branch...

Can you send me the code around line 1138 in your version cellBrowser.js ?
Or send the me the whole file?
Somehow, in my cellBrowser.js in the master branch, it doesn't match what
your error message shows.

Don't you want to make the website publicly accessible? You'll have to do
this anyways sat some point, right? If you want to keep the private, you
can send me the URL via direct email.

**Comment by tkakar** [comment]
@maximilianh. I'm emailing the whole file, but the code around line 1138 is below

```
for (var category of categories) {
                var filtList = filtVals[category];
                var domList = domVals[category];

                let found = false;
                if (filtList.length===0)
                    found = true;
                for (var filtVal of filtList)
                    if (domList.indexOf(filtVal)!=-1)
                        found = true;

                if (!found) {
                    isShown = false;
                    break;
                }
            }
   ```
            
Unfortunately, due to data confidentiality, I cannot share the website.

**Comment by maximilianh** [comment]
Sorry, again a question: which line here is line 1138 ?

**Comment by tkakar** [comment]
It's this line. `if (filtList.length===0)`

**Comment by maximilianh** [comment]
So I'm sure that you got the hierarchy to work, now it's something with the
filters that is broken on your installation. I wonder if you specified some
filters but not others... or the code has a problem (that is not visible on
our site).

OK, I took the entire develop branch and made a pip release from it. Can
you do a "pip install cellbrowser --pre --upgrade" and then run "cbUpgrade
-o <htmlDirectory> --code" to upgrade your javascript files and see if that
helps? If it helps, then this is something that is already fixed somehow in
the develop branch and all we need to do is release the current code.

If it's not fixed, then it would be helpful to see your cellbrowser.conf
files. Or - if you know how to do that - you could put a breakpoint at line
1138 and let me know what the value of the variable "category" is. Or you
could add console.log(category) before line 1138 and send us a copy of the
console.

This will be fixed one way or another, thanks for bearing with me. It's not
easy to debug without seeing the website nor the data.

**Comment by tkakar** [comment]
@maximilianh thanks for troubleshooting this with me. My .conf files were based on some previous version from repo where other categories were not defined (except the body_parts). I just updated it and it's working now. Thanks for all your help.

**Comment by maximilianh** [comment]
Ahhh. This makes sense. So something is broken when these are not
specified. I guess we never tested this.

@Brittney Wick ***@***.***> don't we have a dataset on cells-test where
some of these attributes are missing? Hmmm... I'm surprised this hasn't
come up before...

**Comment by maximilianh** [comment]
Yes, please don't use the -r option, it's not meant to be used with the -p
option, sorry I didn't mention that (and it's not documented I just saw).

No need for the -r just run cbBuild with the -o and -p options.

Also, if this is a production webserver, don't use the -p built-in server,
it's slow and can use a lot of RAM. I'd rather use any other webserver,
apache, nginx or anything else you like.

**Comment by maximilianh** [comment]
Oops, sorry! I should have written "desc.conf", in addition to
"cellbrowser.conf". You define the title in cellbrowser.conf like this
/home/cellbrowser/User1/:

shortLabel = "All Datasets"

and then write your html into desc.conf, as you would do for a normal
dataset:

title = "UCSC Cell Browser Intro"
abstractFile="../news/combined.html"

You can create sample files for the root dataset as with any other dataset:

cd /home/cellbrowser/User1/
cbBuild --init

If you have ideas on how we could better document this or make it easier,
let me know...
